src/polly/src/epoll.rs

```diff
@@ -0,0 +1,201 @@
+        /// event file descriptor via EPOLL_CTL_DEL, or the clearing of EPOLLWAKEUP
+        /// for the event file descriptor via EPOLL_CTL_MOD.
+        const WAKE_UP = EPOLLWAKEUP as u32;
+        /// Sets the one-shot behavior for the associated file descriptor.
+        const ONE_SHOT = EPOLLONESHOT as u32;
+        /// Sets an exclusive wake up mode for the epoll file descriptor that is being
+        /// attached to the associated file descriptor.
+        /// When a wake up event occurs and multiple epoll file descriptors are attached to
+        /// the same target file using this mode, one or more of the epoll file descriptors
+        /// will receive an event with `epoll_wait`. The default here is for all those file
+        /// descriptors to receive an event.
+        const EXCLUSIVE = EPOLLEXCLUSIVE as u32;
+    }
+}
+
+/// This is a wrapper over 'libc::epoll_event'.
+///
+/// We are using `transparent` here to be super sure that this struct and its fields
+/// have the same alignment as those from the `epoll_event` struct from C.
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+pub struct Event(epoll_event);
+
+impl Deref for Event {
+    type Target = epoll_event;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for Event {
+    fn default() -> Self {
+        Event(epoll_event {
+            events: 0u32,
+            u64: 0u64,
+        })
+    }
+}
+
+impl Event {
+    /// Create a new epoll_event instance with the following fields: `events`, which contains
+    /// an event mask and `data` which represents a user data variable. `data` field can be
+    /// a fd on which we want to monitor the events specified by `events`.
+    pub fn new(events: EventType, data: u64) -> Self {
+        Event(epoll_event {
+            events: events.bits(),
+            u64: data,
+        })
+    }
+
+    pub fn events(&self) -> u32 {
+        self.events
+    }
+
+    pub fn data(&self) -> u64 {
+        self.u64
+    }
+}
+
+/// Stores a file descriptor referring to an epoll instance.
+#[derive(Clone, Copy, Debug)]
+pub struct Epoll {
+    epoll_fd: RawFd,
+}
+
+impl Epoll {
+    /// Create a new epoll file descriptor.
+    pub fn new() -> io::Result<Self> {
+        let epoll_fd = SyscallReturnCode(unsafe { epoll_create1(EPOLL_CLOEXEC) }).into_result()?;
+        Ok(Epoll { epoll_fd })
+    }
+
+    /// Wrapper for `libc::epoll_ctl`.
+    ///
+    /// This can be used for adding, modifying or removing a file descriptor in the
+    /// interest list of the epoll instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `operation` refers to the action to be performed on the file descriptor.
+    /// * `fd` is the file descriptor on which we want to perform `operation`.
+    /// * `event` refers to the `epoll_event` instance that is linked to `fd`.
+    pub fn ctl(self, operation: ControlOperation, fd: RawFd, mut event: Event) -> io::Result<()> {
```

Comment:
It is a bit counter-intuitive that an input parameter needs to be mutable. We should either try to find a workaround for this, or add a comment to mention why this needs to be mut. I would prefer the former.

Reply:
I am also curious as why this needs to be mutable.

Reply:
@sandreim  `libc::epoll_ctl()` requires that `epoll_event` parameter to be mutable even if it doesn't actually mutate the object. @andreeaflorescu found a nice workaround for it, take a look!